### PR TITLE
Rely on response content type to render or not the web console

### PR DESCRIPTION
**Pull request by jmbejar**
This PR goal is to improve the logic behind the decision of render or not the web console in the response when some error happens in a Rails app.

We found this issue testing the Rails API-only feature that was recently merged into Rails' master. For instance, we found that the web console code was being appended to JSON responses.

However, it is not only a problem with API-only apps. This issue occurs because `Rails` and `web-console` were using a different logic to calculate the response format (in other words, the response's content type). While `Rails` defines the response format based on the extension in the URL and the Accept headers in the request, `web-console` was relying on the request `Content-Type` header.
Even worse, the response type in Rails can still be overrided by `ActionDispatch::DebugExceptions` to return a response with an HTML response, even if the client requested a different format.

In any case, the request `Content-Type` header should not be used to deduce the response content type. Rails code is not doing that. According [RFC 2616 section 7.2.1](http://www.w3.org/Protocols/rfc2616/rfc2616-sec7.html#sec7.2.1), this header indicates the format of the data in the request body.

In Rails, [this code is the responsible for setting the response content type](https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/http/mime_negotiation.rb#L50-L72).

The approach in this PR is to grab the response content type from the headers that are being passed through the middleware chain. When `web-console` middleware is processed, Rails already returned the response content type so it's easy to check if the response is a html page (where the console can be appended) or something different.

### CODE REVIEW

**Comment by jmbejar** [comment]
@gsamokovarov I'm not 100% happy with part of the resulting code after changes. For example, there is a bit of repetition for the logging part, and maybe I've should put the acceptable content type logic out of the middleware.

However, I would like to hear your thoughts about the overall change before spending more time on refactoring.

**Comment by gsamokovarov** [comment]
Hey @jmbejar, thanks for taking a stab on this! It has been a problem for us and I'm open for any effort to make it better.

> I'm not 100% happy with part of the resulting code after changes. For example, there is a bit of repetition for the logging part, and maybe I've should put the acceptable content type logic out of the middleware.

Let's flesh it out, I'll leave a couple comments on the diff.

File: lib/web_console/middleware.rb [hunk 1]
```diff
@@ -58,6 +59,25 @@ def call(env)
 
     private
 
+      # Returns whether the response is from an acceptable content type.
+      #
+      # We can render a console for HTML and TEXT by default.
+      def acceptable_response_content_type?(headers, env)
```
Comment (gsamokovarov) [inline lib/web_console/middleware.rb]: Did you moved that out of the request object, because you needed the Rack headers? If its that, I think we can keep the method in the request object by making it take the rack headers as input.

Comment (jmbejar) [inline lib/web_console/middleware.rb]: It didn't move it out because I needed the Rack headers, at least it was not the only reason. Since we are checking **response** content type, it does not make sense to include it in the `Request` class. For me, it would lead to confusion between request and response concepts.

Let me try to create a new `Response` class (a wrapper around Rack::Response) adding this specific checking, following the same approach you already did for the requests. I think it would be a nice fit in terms of design.

Comment (gsamokovarov) [inline lib/web_console/middleware.rb]: I get it now. Let's try that. :+1:

**Comment by jmbejar** [comment]
@gsamokovarov You should see now better organized code in this PR :smiley:

Please let me know what are your thoughts.

File: lib/web_console/middleware.rb [hunk 2]
```diff
@@ -123,5 +130,16 @@ def respond_with_unacceptable_request
           { error: UNACCEPTABLE_REQUEST_MESSAGE }
         end
       end
+
+      def log_unless(condition)
```
Comment (gsamokovarov) [inline lib/web_console/middleware.rb]: :scissors:

File: lib/web_console/middleware.rb [hunk 3]
```diff
@@ -123,5 +130,16 @@ def respond_with_unacceptable_request
...
+        unless condition
+          logger.info { yield }
+        end
+        condition
+      end
+
+      def logger
```
Comment (gsamokovarov) [inline lib/web_console/middleware.rb]: :scissors:

File: lib/web_console/middleware.rb [hunk 4]
```diff
@@ -18,8 +18,9 @@ class Middleware
 
     UNACCEPTABLE_REQUEST_MESSAGE = "A supported version is expected in the Accept header."
 
-    cattr_accessor :whiny_requests
-    @@whiny_requests = true
+    cattr_accessor :verbose_logging, :acceptable_content_types
+    @@verbose_logging = true
+    @@acceptable_content_types = [Mime::HTML]
```
Comment (gsamokovarov) [inline lib/web_console/middleware.rb]: This option should be in the Response class, right? Actually, do we need it to be configurable anymore? I never documented it for a reason – every case where a console wasn't shown, I treated like a bug, but gave the users an option to fix their problems in the mean time.

File: lib/web_console/railtie.rb [hunk 5]
```diff
@@ -55,9 +55,9 @@ class Railtie < ::Rails::Railtie
       end
     end
 
-    initializer 'web_console.whiny_requests' do
-      if config.web_console.key?(:whiny_requests)
-        Middleware.whiny_requests = config.web_console.whiny_requests
+    initializer 'web_console.verbose_logging' do
+      if config.web_console.key?(:verbose_logging)
```
Comment (gsamokovarov) [inline lib/web_console/railtie.rb]: We can't just rename that, its already documented [here](https://github.com/rails/web-console#configweb_consolewhiny_requests). Even if its not _just_ for requests anymore, we don't wanna break any applications using the previous option.

File: lib/web_console/response.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
+module WebConsole
+  # Web Console tailored response object.
+  class Response < Rack::Response
+
```
Comment (gsamokovarov) [inline lib/web_console/response.rb]: :scissors:

File: lib/web_console/response.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+      content_type == Mime::HTML
+    end
+
+    def content_type
+      formats = Mime::Type.parse(headers['Content-Type'])
+      formats.first
+    end
+  end
+end
+
```
Comment (gsamokovarov) [inline lib/web_console/response.rb]: :scissors:

File: lib/web_console/whiny_response.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
+module WebConsole
+  # Noisy wrapper around +Response+.
+  #
+  # If any calls to +acceptable_content_type?+ return false,
+  # an info log message will be displayed in users' logs.
+  class WhinyResponse < WhinyObject
+
```
Comment (gsamokovarov) [inline lib/web_console/whiny_response.rb]: :scissors:

File: lib/web_console/whiny_object.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
+module WebConsole
+  class WhinyObject < SimpleDelegator
+    def initialize(object, env)
+      @env = env
+      super(object)
+    end
+
+    private
+
+      attr_accessor :env
```
Comment (gsamokovarov) [inline lib/web_console/whiny_object.rb]: You don't need that attr.

File: lib/web_console/whiny_object.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
+module WebConsole
+  class WhinyObject < SimpleDelegator
+    def initialize(object, env)
```
Comment (gsamokovarov) [inline lib/web_console/whiny_object.rb]: No need for the env.

File: lib/web_console/whiny_request.rb [hunk 11]
```diff
@@ -1,9 +1,10 @@
 module WebConsole
   # Noisy wrapper around +Request+.
   #
-  # If any calls to +from_whitelisted_ip?+ and +acceptable_content_type?+
-  # return false, an info log message will be displayed in users' logs.
-  class WhinyRequest < SimpleDelegator
+  # If any calls to +from_whitelisted_ip?+ return false,
+  # an info log message will be displayed in users' logs.
+  class WhinyRequest < WhinyObject
+
```
Comment (gsamokovarov) [inline lib/web_console/whiny_request.rb]: :scissors:

File: lib/web_console/whiny_request.rb [hunk 12]
```diff
@@ -1,9 +1,10 @@
 module WebConsole
   # Noisy wrapper around +Request+.
   #
-  # If any calls to +from_whitelisted_ip?+ and +acceptable_content_type?+
-  # return false, an info log message will be displayed in users' logs.
-  class WhinyRequest < SimpleDelegator
+  # If any calls to +from_whitelisted_ip?+ return false,
```
Comment (gsamokovarov) [inline lib/web_console/whiny_request.rb]: You sure this is wrapped by 80 chars as before and not something lower?

File: lib/web_console/whiny_object.rb [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+      @env = env
+      super(object)
+    end
+
+    private
+
+      attr_accessor :env
+
+      def whine_unless(condition)
```
Comment (gsamokovarov) [inline lib/web_console/whiny_object.rb]: I think `log_unless` was a bit clearer. Changing method names because you find them clearer is kinda subjective and just contributes to the diff noise. Think of it as a cosmetic change. Those aren't usually accepted across the rails core projects.

File: test/web_console/helper_test.rb [hunk 14]
```diff
@@ -23,6 +23,14 @@ def call(env)
         def request
           Request.new(@env)
         end
+
+        def status
```
Comment (gsamokovarov) [inline test/web_console/helper_test.rb]: Do the tests pass without the status and the header? The console can be shown on regular pages as well. If we don _really_ need the status and the headers, I would cut them just so we have smaller diff.

File: lib/web_console/middleware.rb (hunk 4 above)
Comment (jmbejar) [inline lib/web_console/middleware.rb]: I'm not longer using the `acceptable_content_types` config setting because I guessed the same thing you're saying: it was an undocumented option to handle buggy cases.

I forgot to remove this line, so let's :scissors:

File: lib/web_console/railtie.rb (hunk 5 above)
Comment (jmbejar) [inline lib/web_console/railtie.rb]: I thought it was not documented like `acceptable_content_types `, but I was wrong. Although I don't like the name, we shouldn't rename it without previous advise to users.

Will revert this change

File: lib/web_console/whiny_object.rb (hunk 13 above)
Comment (jmbejar) [inline lib/web_console/whiny_object.rb]: Ops, it was changed by mistake. I also prefer `log_unless` so it's all right.

File: lib/web_console/whiny_object.rb [hunk 15]
```diff
@@ -0,0 +1,23 @@
+module WebConsole
+  class WhinyObject < SimpleDelegator
```
Comment (gsamokovarov) [inline lib/web_console/whiny_object.rb]: I have been thinking about this for a while: it may be better if we drop the decorators and use `ActiveSupport::Notifications.instrument` for logging. This can happen in the middleware and we can create a subscriber that logs the information if the users set `whiny_requests` to true.

This will drop the decorators as a whole. It didn't matter that much when we had just one object that we logged, but now that we have more I don't think the extra overhead is worth it. @jmbejar WDYT?

File: lib/web_console/whiny_object.rb (hunk 13 above)
Comment (jmbejar) [inline lib/web_console/whiny_object.rb]: @gsamokovarov Actually, I'm a bit confused :confused: 

`whine_unless` is the method name that is currently on [master](https://github.com/rails/web-console/blob/master/lib/web_console/whiny_request.rb#L23)

`log_unless` was the method name I used when the logic was in the middleware, so it's actually was introduced by me.

If understand well your comment about avoiding cosmetic changes, we should use `whine_unless` (so, for example, we wouldn't have to change this line https://github.com/rails/web-console/blob/master/lib/web_console/whiny_request.rb#L8)

File: test/web_console/helper_test.rb (hunk 14 above)
Comment (jmbejar) [inline test/web_console/helper_test.rb]: Unfortunately, tests do not pass without the header because it's getting parsed by the logic that checks if the console has to be included.

Regarding the status, we're forced to pass it because is the second arg. We _could_ pass a nil value, but I would prefer to use some valid value (e.g. 500, 200, etc) because having incomplete or invalid test data could lead to confusion later.

Anyway, I'm open to change the invocation if you want to use another status value or just pass nil.

File: lib/web_console/whiny_object.rb (hunk 15 above)
Comment (jmbejar) [inline lib/web_console/whiny_object.rb]: Since I was not familiar with the `ActiveSupport::Notifications` API, I just decided to play a bit in order to form some opinion.

I think you are thinking on using `ActiveSupport::LogSubscriber` and trigger events with `ActiveSupport::Notifications.instrument`.

I tested all this stuff with changes in the `middleware.rb` file. A working (but still ugly) solution is [here](https://gist.github.com/jmbejar/95854eb3005a78b915c4).

I personally feel this is too much overhead to just render a log message. In fact, looking at the Rails code, this is used just in very few occasions (sql logs, time to render the view). Looks like `ActiveSupport::Notifications` is used when there is the need to really instrument something (I mean, measuring time spent and so on). For this reason, the only way to trigger events is by calling `instrument` method (as far I can say).

Also, we still have to decide where the conditional logic to show the log messages goes, the `ActiveSupport::Notifications` would be just a fancy way to call the logger.

Regarding where the conditional to show logs goes, I see 3 options:

- Have it on the middleware (it was my first approach)
- Have specific classes to log these errors (this was your initial approach and it was I followed)
- Have the logic inside Response and Request (logger can be retrieved from the env in the middleware and passed to these objects in initialization)

I recognize that having 3 `whipy` classes can be a bit of overhead in this case, however I wouldn't go with the first option. The second option seems feasible and I could try it if you want.

Comment (jmbejar) [inline lib/web_console/whiny_object.rb]: I finally put hands on the third option, so I added a new commit implementing that: https://github.com/rails/web-console/commit/60eb494473a110af7bb6b235a7a84efb20d77b3a

I really like the result, no more specific classes doing only the log decision part. The responsibility went to Response and Request classes, with a little part into the Middleware that is taking care of initializing the logger object and passing along to the other collaborators.

File: lib/web_console/whiny_object.rb (hunk 13 above)
Comment (jmbejar) [inline lib/web_console/whiny_object.rb]: If you buy this commit https://github.com/rails/web-console/commit/60eb494473a110af7bb6b235a7a84efb20d77b3a, I think this discussion is not relevant now.

File: lib/web_console/whiny_object.rb (hunk 15 above)
Comment (jmbejar) [inline lib/web_console/whiny_object.rb]: @gsamokovarov Although I'm satisfied with the last approach, I'm looking forward to hear from you what is your preference.

Comment (gsamokovarov) [inline lib/web_console/whiny_object.rb]: @jmbejar So I got confused here, sorry for that. Dunno why I called that class `Whiny` something to begin with :sweat_smile: I kinda like the logging separation, because I don't like having the request/response objects to log messages, I want most of the methods do be simple queries.

However, this is getting to be a but of an overkill for more than one classes, so I'm okay with your current approach.

I can experiment with separated logging on a separate PR, so no need to bother you with it. I'll take a final look over the PR, but it looks better now.

Thank you for taking the time to work on this.

File: lib/web_console/middleware.rb [hunk 16]
```diff
@@ -73,9 +78,12 @@ def json_response_with_session(id, request, opts = {})
         json_response(opts) { yield session }
       end
 
-      def create_regular_or_whiny_request(env)
-        request = Request.new(env)
-        whiny_requests ? WhinyRequest.new(request) : request
+      def create_request(env, logger)
```
Comment (gsamokovarov) [inline lib/web_console/middleware.rb]: You don't need those two methods anymore.

File: lib/web_console/middleware.rb [hunk 17]
```diff
@@ -44,7 +49,7 @@ def call(env)
         session = Session.from_binding(binding)
       end
 
-      if session && request.acceptable_content_type?
+      if session && response.acceptable_content_type?
         headers["X-Web-Console-Session-Id"] = session.id
         response = Rack::Response.new(body, status, headers)
```
Comment (gsamokovarov) [inline lib/web_console/middleware.rb]: You can reuse the already instantiated `WebConsole::Response` here.

File: lib/web_console/request.rb [hunk 18]
```diff
@@ -26,20 +31,22 @@ def strict_remote_ip
...
-    end
-
     # Returns whether the request is acceptable.
     def acceptable?
       xhr? && accepts.any? { |mime| Mime::WEB_CONSOLE_V2 == mime }
     end
 
+    private
+
+    attr_reader :logger
```
Comment (gsamokovarov) [inline lib/web_console/request.rb]: Can you indent the private block here?

File: lib/web_console/middleware.rb (hunk 17 above)
Comment (jmbejar) [inline lib/web_console/middleware.rb]: Good catch!

Already changed among with the rest of your comments

**Comment by gsamokovarov** [comment]
@jmbejar Can you also rebase and squish the commits?

**Comment by jmbejar** [comment]
@gsamokovarov Commits are rebased and squashed now.

Thanks for all your help and guidance! :fireworks:

File: test/web_console/middleware_test.rb [hunk 19]
```diff
@@ -15,6 +19,16 @@ def call(env)
           </html>
         HTML
       end
+
+      private
+
+      def status
```
Comment (sh19910711) [inline test/web_console/middleware_test.rb]: Indent under private?

Comment (jmbejar) [inline test/web_console/middleware_test.rb]: Good one! Fixing now (by amending the commit)

**Comment by jmbejar** [comment]
@gsamokovarov Let me know if we need something else for this PR.

This https://github.com/rails/rails/pull/20831 is depending on this.

**Comment by gsamokovarov** [comment]
Thanks!

**Comment by jmbejar** [comment]
:+1:
